In [59]:
data(Boston,package="MASS")

In [60]:
Boston$medv <- NULL
Boston$chas <- NULL
Boston$zn <- NULL
dim(Boston)

[1] 506  11

In [61]:
str(Boston)

'data.frame':	506 obs. of  11 variables:
 $ crim   : num  0.00632 0.02731 0.02729 0.03237 0.06905 ...
 $ indus  : num  2.31 7.07 7.07 2.18 2.18 2.18 7.87 7.87 7.87 7.87 ...
 $ nox    : num  0.538 0.469 0.469 0.458 0.458 0.458 0.524 0.524 0.524 0.524 ...
 $ rm     : num  6.58 6.42 7.18 7 7.15 ...
 $ age    : num  65.2 78.9 61.1 45.8 54.2 58.7 66.6 96.1 100 85.9 ...
 $ dis    : num  4.09 4.97 4.97 6.06 6.06 ...
 $ rad    : int  1 2 2 3 3 3 5 5 5 5 ...
 $ tax    : num  296 242 242 222 222 222 311 311 311 311 ...
 $ ptratio: num  15.3 17.8 17.8 18.7 18.7 18.7 15.2 15.2 15.2 15.2 ...
 $ black  : num  397 397 393 395 397 ...
 $ lstat  : num  4.98 9.14 4.03 2.94 5.33 ...


In [62]:
Boston_log_df = log(Boston)
Boston_log_sc_df = data.frame(scale(Boston_log_df))

In [63]:
display_outliers <- function (dataframe, feature, param=1.5) {
    feature_vec =  as.vector(dataframe[[feature]])
    Q1 <- quantile(feature_vec, .25)
    Q3 <- quantile(feature_vec, .75)
    tukey_window <- param*(Q3-Q1)
    less_than_Q1 <- dataframe[[feature]] < Q1 - tukey_window
    greater_than_Q3 <- dataframe[[feature]] > Q3 + tukey_window
    tukey_mask <- (less_than_Q1 | greater_than_Q3)
    return(dataframe[tukey_mask,])
}

In [64]:
display_outliers(Boston_log_sc_df, 'crim')

crim indus nox rm age dis rad tax ptratio black lstat

In [65]:
display_outliers(Boston_log_sc_df, 'indus')

crim      indus     nox       rm        age        dis      rad       
57  -1.435886 -3.167744 -1.397500 0.1938494 -0.8443067 1.908681 -1.3425573
196 -1.619725 -3.779625 -1.254321 2.0639136 -1.0234577 1.007031 -0.5502381
    tax        ptratio    black     lstat    
57  -0.4672479 -0.4583125 0.3022931 -1.028960
196 -0.9842937 -1.9282409 0.2935664 -2.134168

In [66]:
display_outliers(Boston_log_sc_df, 'nox')

crim indus nox rm age dis rad tax ptratio black lstat

In [67]:
display_outliers(Boston_log_sc_df, 'rm')

crim        indus      nox        rm        age        dis        
98  -0.61651407 -1.4143540 -0.9909291  2.280574  0.3928518  0.11742862
145  0.83383522  1.0480441  2.3421958 -2.154612  0.8057788 -1.65133018
148  0.75980831  1.0480441  2.3421958 -2.112947  0.7702379 -1.49949657
164  0.55433570  1.0480441  0.5335386  2.611947  0.7391479 -0.77286880
167  0.68391819  1.0480441  0.5335386  2.124753  0.7787703 -0.87516939
205 -1.44635713 -1.5114466 -1.3242013  2.241874 -1.0285825  0.82426875
225 -0.17284487 -0.4319792 -0.3730008  2.495318  0.4416684 -0.23214784
226  0.06463712 -0.4319792 -0.3730008  2.976438  0.5371148 -0.23214784
227 -0.08396289 -0.4319792 -0.3730008  2.248520  0.6047439 -0.03704469
233  0.10524975 -0.4319792 -0.3435455  2.571461  0.3336241  0.29102813
234 -0.14975679 -0.4319792 -0.3435455  2.474831  0.2675283  0.19871351
254 -0.10022199 -0.5045669 -1.1495837  2.487776 -3.2134294  1.85113119
258  0.13351278 -1.0057121  0.8666578  2.954984  0.6122980 -1.11147102
263  0.05863832 -1.0057121  0.8666578  2.636363  0.6967544 -0.66747870
268  0.10769542 -1.0057121  0.2811181  2.528644  0.1864781 -0.56270201
284 -1.58118542 -2.5348821 -1.5076620  2.118013 -1.4408073  1.08308425
365  0.93699187  0.9468884  1.3834425  3.032382  0.5351409 -1.00771269
366  1.06234938  0.9468884  1.3834425 -5.001759  0.6310323 -1.31557257
368  1.56553670  0.9468884  0.7423772 -4.277053  0.8422028 -1.43736533
374  1.47457757  0.9468884  1.0251918 -2.149166  0.8422028 -1.90427568
375  1.71046425  0.9468884  1.0251918 -3.664825  0.8422028 -1.96394403
385  1.74852759  0.9468884  1.2574306 -3.183252  0.6913772 -1.52672017
387  1.83842407  0.9468884  1.2574306 -2.622451  0.8422028 -1.49139422
389  1.59249288  0.9468884  1.2574306 -2.196473  0.8422028 -1.34300354
407  1.76284165  0.9468884  0.9578678 -3.664825  0.8422028 -1.89812995
413  1.71822002  0.9468884  0.4674717 -2.668500  0.8422028 -1.38498620
415  2.12924873  0.9468884  1.2075488 -2.880688  0.8422028 -1.26457995
    rad         tax        ptratio    black      lstat     
98  -1.34255729 -0.7846368 -0.1405361  0.3022931 -1.5535296
145 -0.29516833  0.1703775 -1.7630504  0.3022931  1.6746472
148 -0.29516833  0.1703775 -1.7630504  0.2852756  1.6882279
164 -0.29516833  0.1703775 -1.7630504  0.2744706 -1.9487723
167 -0.29516833  0.1703775 -1.7630504  0.2091093 -1.7684268
205 -0.55023813 -1.3113079 -1.7630504  0.2814412 -2.1853782
225  0.24208104 -0.5160801 -0.4121369  0.2631047 -1.5814330
226  0.24208104 -0.5160801 -0.4121369  0.2528230 -1.3952744
227  0.24208104 -0.5160801 -0.4121369  0.2709045 -2.0468462
233  0.24208104 -0.5160801 -0.4121369  0.2659890 -2.4409521
234  0.24208104 -0.5160801 -0.4121369  0.2424589 -1.6596172
254  0.08944465 -0.3338122  0.3346761  0.3022931 -1.8419943
258 -0.29516833 -0.8967849 -2.7476435  0.2786243 -1.2278605
263 -0.29516833 -0.8967849 -2.7476435  0.2691678 -0.9890629
268 -0.29516833 -0.8967849 -2.7476435  0.2613911 -0.6059237
284 -2.13487646 -1.6225830 -2.3861631  0.2977901 -2.0309714
365  1.49787721  1.4377728  0.7832718  0.1564118 -1.1735015
366  1.49787721  1.4377728  0.7832718  0.1569587 -0.6790869
368  1.49787721  1.4377728  0.7832718 -1.1267015  0.3645451
374  1.49787721  1.4377728  0.7832718  0.3022931  1.9600711
375  1.49787721  1.4377728  0.7832718  0.3022931  2.1065892
385  1.49787721  1.4377728  0.7832718 -0.1221405  1.7490928
387  1.49787721  1.4377728  0.7832718  0.3022931  1.6162484
389  1.49787721  1.4377728  0.7832718  0.2217207  1.7485494
407  1.49787721  1.4377728  0.7832718  0.2123261  1.2967460
413  1.49787721  1.4377728  0.7832718 -3.0897615  1.9408150
415  1.49787721  1.4377728  0.7832718 -1.6412624  2.0626226

In [68]:
display_outliers(Boston_log_sc_df, 'age')

crim       indus      nox        rm            age       dis      
42  -0.5918796 -0.2924398 -0.9575816  0.7178901697 -4.954793 1.0306687
43  -0.5434745 -0.2924398 -0.9575816 -0.1097468319 -3.608298 1.0306687
44  -0.4884961 -0.2924398 -0.9575816 -0.0493403099 -3.633296 1.0306687
71  -0.7617917  0.2835052 -1.3613162  0.2411453478 -3.608298 0.8845860
73  -0.7444097  0.2835052 -1.3613162 -0.2611130097 -3.334771 0.8845860
74  -0.3942191  0.2835052 -1.3613162 -0.0007382233 -3.710666 0.8845860
75  -0.8132917  0.5039909 -1.0809669  0.0390887523 -3.764355 0.4804766
194 -1.4070918 -1.3966627 -1.5076620  0.7572538443 -2.944407 1.1855756
201 -1.5028535 -2.2843735 -1.4829694  1.1853832307 -2.388763 1.5700548
215 -0.2122943  0.2570422 -0.5229578 -1.2752756559 -2.961030 0.1657377
244 -0.5914080 -0.7269780 -1.1842511  0.2077861085 -3.334771 1.4140326
251 -0.5474137 -0.5045669 -1.1495837  0.3377351919 -2.498367 1.5068238
252 -0.3519449 -0.5045669 -1.1495837  0.2702324944 -3.118758 1.5068238
253 -0.7946356 -0.5045669 -1.1495837  0.9604653483 -3.559418 1.8511312
254 -0.1002220 -0.5045669 -1.1495837  2.4877758182 -3.213429 1.8511312
300 -0.9754426 -1.7422624 -1.5200545  1.0673147340 -2.927951 1.6118148
326 -0.4026516 -0.2077486 -0.4825242  0.2605473344 -2.297139 0.9291259
    rad         tax        ptratio    black     lstat     
42  -0.87908029 -1.2119242 -0.1851683 0.2643129 -1.3214543
43  -0.87908029 -1.2119242 -0.1851683 0.2574514 -1.0174628
44  -0.87908029 -1.2119242 -0.1851683 0.2943205 -0.6059237
71  -0.55023813 -0.5325698  0.3765115 0.2586649 -0.7753099
73  -0.55023813 -0.5325698  0.3765115 0.2826324 -1.1026741
74  -0.55023813 -0.5325698  0.3765115 0.2363717 -0.5837045
75  -0.29516833  0.1388800  0.1651151 0.2958273 -0.7605170
194 -2.13487646 -0.8872465 -1.2869823 0.2907430 -1.2573740
201 -0.87908029  0.1641093 -0.5984582 0.2605839 -1.4612643
215 -0.55023813 -0.7755123  0.1221582 0.1357542  1.6893546
244 -0.08676112 -0.5742719 -0.7892162 0.2279116 -1.2052619
251  0.08944465 -0.3338122  0.3346761 0.3002720 -0.9918812
252  0.08944465 -0.3338122  0.3346761 0.2360289 -1.8186532
253  0.08944465 -0.3338122  0.3346761 0.2665919 -1.8467021
254  0.08944465 -0.3338122  0.3346761 0.3022931 -1.8419943
300 -0.29516833 -0.1283450 -1.7087352 0.2170667 -1.3561986
326 -0.29516833 -0.6860378  0.5417020 0.2917614 -1.2409130

In [69]:
display_outliers(Boston_log_sc_df, 'dis')

crim indus nox rm age dis rad tax ptratio black lstat

In [70]:
display_outliers(Boston_log_sc_df, 'rad')

crim indus nox rm age dis rad tax ptratio black lstat

In [71]:
display_outliers(Boston_log_sc_df, 'tax')

crim indus nox rm age dis rad tax ptratio black lstat

In [72]:
display_outliers(Boston_log_sc_df, 'ptratio')

crim        indus     nox        rm         age         dis       
197 -1.12656647 -2.241325 -1.4706690  1.3730501 -0.91938489  1.4847173
198 -1.05660436 -2.241325 -1.4706690  1.1503773 -0.80354049  1.4847173
199 -1.15547247 -2.241325 -1.4706690  1.3571535 -0.72920178  1.4847173
258  0.13351278 -1.005712  0.8666578  2.9549841  0.61229802 -1.1114710
259  0.17123788 -1.005712  0.8666578  1.4290730  0.84220282 -1.0175669
260  0.16643098 -1.005712  0.8666578  0.8120722  0.84220282 -0.9073351
261  0.07606382 -1.005712  0.8666578  1.2698288  0.51326938 -0.8161477
262  0.07090562 -1.005712  0.8666578  1.6532566  0.65873780 -0.7919985
263  0.05863832 -1.005712  0.8666578  2.6363631  0.69675442 -0.6674787
264  0.27213959 -1.005712  0.8666578  1.4217856  0.74957697 -0.8456019
265  0.08451539 -1.005712  0.8666578  1.2735360  0.69854291 -0.9831600
266  0.23502170 -1.005712  0.8666578 -1.0350852  0.08047966 -0.9297773
267  0.24941882 -1.005712  0.8666578  1.0331099  0.56837794 -0.7979846
268  0.10769542 -1.005712  0.2811181  2.5286437  0.18647813 -0.5627020
269  0.07639768 -1.005712  0.2811181  1.5938653 -0.20972375 -0.2465473
284 -1.58118542 -2.534882 -1.5076620  2.1180132 -1.44080729  1.0830842
    rad        tax        ptratio   black     lstat      
197 -1.3425573 -0.3414690 -2.998022 0.3022931 -1.60572820
198 -1.3425573 -0.3414690 -2.998022 0.1555364 -0.36286235
199 -1.3425573 -0.3414690 -2.998022 0.2868919 -0.80026081
258 -0.2951683 -0.8967849 -2.747643 0.2786243 -1.22786046
259 -0.2951683 -0.8967849 -2.747643 0.2571816 -0.52942068
260 -0.2951683 -0.8967849 -2.747643 0.2860015 -0.73131983
261 -0.2951683 -0.8967849 -2.747643 0.2888682 -0.18346791
262 -0.2951683 -0.8967849 -2.747643 0.2742044 -0.64668153
263 -0.2951683 -0.8967849 -2.747643 0.2691678 -0.98906290
264 -0.2951683 -0.8967849 -2.747643 0.2909073  0.08221613
265 -0.2951683 -0.8967849 -2.747643 0.2726055 -0.46447849
266 -0.2951683 -0.8967849 -2.747643 0.2875510 -0.04054508
267 -0.2951683 -0.8967849 -2.747643 0.2598099  0.53751167
268 -0.2951683 -0.8967849 -2.747643 0.2613911 -0.60592372
269 -0.2951683 -0.8967849 -2.747643 0.2806134 -2.03097140
284 -2.1348765 -1.6225830 -2.386163 0.2977901 -2.03097140

In [73]:
display_outliers(Boston_log_sc_df, 'black')

crim       indus       nox         rm           age          dis       
19   0.2593254 -0.08159911 -0.04899171 -1.203188326 -0.8035405    0.2706851
26   0.2806251 -0.08159911 -0.04899171 -0.972855925  0.5895302    0.5669666
28   0.3400468 -0.08159911 -0.04899171 -0.287574267  0.6477118    0.5664673
33   0.5126120 -0.08159911 -0.04899171 -0.431540970  0.5172678    0.3628210
35   0.5820496 -0.08159911 -0.04899171 -0.215724401  0.7906414    0.2526814
103 -0.3212902 -0.01684901 -0.21788816  0.224481353  0.5837885   -0.3509446
116 -0.4549705  0.18454990  0.03334898 -0.464519658  0.6366110   -0.5312085
119 -0.5806215  0.18454990  0.03334898 -0.549020843  0.3291504   -0.5204044
135  0.3498151  1.19157445  0.68701021 -0.725106162  0.8157933   -0.6214861
146  0.7618969  1.04804409  2.34219579 -0.166207990  0.8422028   -1.5531738
147  0.7161028  1.04804409  2.34219579 -0.926863241  0.8422028   -1.4300183
152  0.5473716  1.04804409  2.34219579 -1.288445387  0.8422028   -1.3405565
153  0.4160971  1.04804409  2.34219579 -1.958860473  0.6328940   -1.3190225
154  0.7148412  1.04804409  2.34219579 -0.799645470  0.8174564   -1.3041190
155  0.5211501  1.04804409  2.34219579 -0.167660431  0.7753626   -1.1653481
156  0.9450069  1.04804409  2.34219579 -0.134314094  0.5292049   -1.1694846
157  0.7748054  1.04804409  2.34219579 -1.508606705  0.7408907   -1.1791724
161  0.4727799  1.04804409  0.53353858  0.006386816  0.7163211   -1.1141486
166  0.8572367  1.04804409  0.53353858 -0.208425281  0.7233787   -0.6716137
168  0.6329078  1.04804409  0.53353858 -0.541443401  0.4603812   -0.5594139
169  0.7462913  1.04804409  0.53353858  0.104134449  0.7770673   -0.8267963
170  0.7753439  1.04804409  0.53353858  0.220310476  0.7616608   -0.6886561
171  0.4481497  1.04804409  0.53353858 -0.544473604  0.7513087   -0.5594139
287 -1.4565997 -2.05264330 -1.70975403 -0.022147620 -1.0492434    1.8887240
367  0.9657243  0.94688838  1.38344248 -2.046326659  0.6949640   -1.1622782
368  1.5655367  0.94688838  0.74237716 -4.277052541  0.8422028   -1.4373653
385  1.7485276  0.94688838  1.25743063 -3.183252043  0.6913772   -1.5267202
400  1.4220952  0.94688838  1.20754880 -0.579395273  0.4311792   -1.4499225
405  2.0845182  0.94688838  1.20754880 -1.081641841  0.5837885   -1.3222482
408  1.5084106  0.94688838  0.95786775 -0.958556890  0.8422028   -1.7368623
⋮   ⋮          ⋮           ⋮           ⋮            ⋮            ⋮         
427  1.5197304 0.9468884   0.3582013   -0.60224430  -0.002408339 -0.9194498
428  2.0393075 0.9468884   1.1062555   -0.06225011   0.450011593 -1.0488400
429  1.2846425 0.9468884   1.1062555   -0.07517865   0.437480739 -0.9778861
430  1.3943356 0.9468884   1.1062555    0.18966419   0.768526073 -0.9469303
431  1.3503738 0.9468884   0.3582013    0.14489861   0.597154723 -0.8690194
432  1.4288438 0.9468884   0.3582013    0.80035380   0.746107988 -0.8372400
433  1.2227253 0.9468884   0.3582013    0.25223737   0.366792488 -0.7402387
434  1.1562251 0.9468884   1.3487589    0.26746638   0.631032280 -0.6454998
435  1.5787277 0.9468884   1.3487589   -0.05364150   0.758217391 -0.7219668
436  1.4767501 0.9468884   1.5332349    0.53051317   0.751308702 -0.8051239
437  1.5952949 0.9468884   1.5332349    0.30198118   0.728651992 -0.9148165
438  1.6189404 0.9468884   1.5332349   -0.13431409   0.842202818 -0.9984915
439  1.5708387 0.9468884   1.5332349   -0.45401318   0.631032280 -1.0914096
445  1.5402330 0.9468884   1.5332349   -0.57635316   0.785564279 -1.0165889
446  1.4560443 0.9468884   1.5332349    0.29922492   0.754766690 -0.9284715
447  1.2113920 0.9468884   1.5332349    0.13507606   0.782170792 -0.8516746
450  1.2945126 0.9468884   1.3487589    0.24114535   0.814128438 -0.7532558
451  1.2419624 0.9468884   1.3487589    0.69023167   0.716321113 -0.6392677
455  1.4029099 0.9468884   1.3487589    0.66248698   0.742631645 -0.5065418
456  1.0818803 0.9468884   1.3487589    0.38973411   0.604743876 -0.5518656
457  1.0736775 0.9468884   1.3487589   -0.39272301  

In [74]:
display_outliers(Boston_log_sc_df, 'lstat')

crim      indus    nox       rm       age     dis        rad       
162 0.5370696 1.048044 0.5335386 1.616481 0.68418 -0.9443896 -0.2951683
    tax       ptratio  black     lstat    
162 0.1703775 -1.76305 0.2269452 -3.033566

In [75]:
summary(Boston)

      crim              indus            nox               rm       
 Min.   : 0.00632   Min.   : 0.46   Min.   :0.3850   Min.   :3.561  
 1st Qu.: 0.08204   1st Qu.: 5.19   1st Qu.:0.4490   1st Qu.:5.886  
 Median : 0.25651   Median : 9.69   Median :0.5380   Median :6.208  
 Mean   : 3.61352   Mean   :11.14   Mean   :0.5547   Mean   :6.285  
 3rd Qu.: 3.67708   3rd Qu.:18.10   3rd Qu.:0.6240   3rd Qu.:6.623  
 Max.   :88.97620   Max.   :27.74   Max.   :0.8710   Max.   :8.780  
      age              dis              rad              tax       
 Min.   :  2.90   Min.   : 1.130   Min.   : 1.000   Min.   :187.0  
 1st Qu.: 45.02   1st Qu.: 2.100   1st Qu.: 4.000   1st Qu.:279.0  
 Median : 77.50   Median : 3.207   Median : 5.000   Median :330.0  
 Mean   : 68.57   Mean   : 3.795   Mean   : 9.549   Mean   :408.2  
 3rd Qu.: 94.08   3rd Qu.: 5.188   3rd Qu.:24.000   3rd Qu.:666.0  
 Max.   :100.00   Max.   :12.127   Max.   :24.000   Max.   :711.0  
    ptratio          black            lst

In [76]:
for (feature in colnames(Boston_log_sc_df)){
    outlier_count = dim(display_outliers(Boston_log_sc_df, feature))[1]
    print(paste(feature, outlier_count))
}

[1] "crim 0"
[1] "indus 2"
[1] "nox 0"
[1] "rm 27"
[1] "age 17"
[1] "dis 0"
[1] "rad 0"
[1] "tax 0"
[1] "ptratio 16"
[1] "black 78"
[1] "lstat 1"


In [77]:
raw_outliers = c()
for (feature in colnames(Boston_log_sc_df)){
    outlier_df = display_outliers(Boston_log_sc_df, feature)
    outlier_indices = rownames(outlier_df)
    raw_outliers = c(raw_outliers, outlier_indices)
}
raw_outliers

[1] "57"  "196" "98"  "145" "148" "164" "167" "205" "225" "226" "227" "233"
 [13] "234" "254" "258" "263" "268" "284" "365" "366" "368" "374" "375" "385"
 [25] "387" "389" "407" "413" "415" "42"  "43"  "44"  "71"  "73"  "74"  "75" 
 [37] "194" "201" "215" "244" "251" "252" "253" "254" "300" "326" "197" "198"
 [49] "199" "258" "259" "260" "261" "262" "263" "264" "265" "266" "267" "268"
 [61] "269" "284" "19"  "26"  "28"  "33"  "35"  "103" "116" "119" "135" "146"
 [73] "147" "152" "153" "154" "155" "156" "157" "161" "166" "168" "169" "170"
 [85] "171" "287" "367" "368" "385" "400" "405" "408" "409" "410" "411" "412"
 [97] "413" "414" "415" "416" "417" "418" "419" "420" "421" "422" "423" "424"
[109] "425" "426" "427" "428" "429" "430" "431" "432" "433" "434" "435" "436"
[121] "437" "438" "439" "445" "446" "447" "450" "451" "455" "456" "457" "458"
[133] "459" "461" "466" "467" "468" "476" "490" "491" "162"

In [78]:
table(raw_outliers)

raw_outliers
103 116 119 135 145 146 147 148 152 153 154 155 156 157 161 162 164 166 167 168 
  1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
169 170 171  19 194 196 197 198 199 201 205 215 225 226 227 233 234 244 251 252 
  1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
253 254 258 259  26 260 261 262 263 264 265 266 267 268 269  28 284 287 300 326 
  1   2   2   1   1   1   1   1   2   1   1   1   1   2   1   1   2   1   1   1 
 33  35 365 366 367 368 374 375 385 387 389 400 405 407 408 409 410 411 412 413 
  1   1   1   1   1   2   1   1   2   1   1   1   1   1   1   1   1   1   1   2 
414 415 416 417 418 419  42 420 421 422 423 424 425 426 427 428 429  43 430 431 
  1   2   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
432 433 434 435 436 437 438 439  44 445 446 447 450 451 455 456 457 458 459 461 
  1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 
466 467 468 476

In [79]:
dim(Boston_log_sc_df)

[1] 506  11

**one outlier**

In [93]:
outlier_counts <- as.data.frame(table(raw_outliers))
outlier_counts

raw_outliers Freq
1   103          1   
2   116          1   
3   119          1   
4   135          1   
5   145          1   
6   146          1   
7   147          1   
8   148          1   
9   152          1   
10  153          1   
11  154          1   
12  155          1   
13  156          1   
14  157          1   
15  161          1   
16  162          1   
17  164          1   
18  166          1   
19  167          1   
20  168          1   
21  169          1   
22  170          1   
23  171          1   
24  19           1   
25  194          1   
26  196          1   
27  197          1   
28  198          1   
29  199          1   
30  201          1   
⋮   ⋮            ⋮   
103 434          1   
104 435          1   
105 436          1   
106 437          1   
107 438          1   
108 439          1   
109 44           1   
110 445          1   
111 446          1   
112 447          1   
113 450          1   
114 451          1   
115 455          1   
116 456          1   
117 457          1   
118 458          1   
119 459          1   
120 461          1   
121 466          1   
122 467          1   
123 468          1   
124 476          1   
125 490          1   
126 491          1   
127 57           1   
128 71           1   
129 73           1   
130 74           1   
131 75           1   
132 98           1

In [87]:
sum(outlier_counts$Freq == 1)

[1] 123

In [89]:
one_outlier=123/506
one_outlier

[1] 0.243083

24.31% of the dataset has outliers in one feature

**two outliers**


In [90]:
sum(outlier_counts$Freq == 2)

[1] 9

In [92]:
two_outlier=9/506
two_outlier

[1] 0.01778656

1.78% of the dataset has outliers in two features

**Plans for handling outliers**

We could perhaps throw out the rows with two outliers as they are only a small part of the dataset. Other ideas include doing capping. For example we can look at the values of the 5th and 95th percentile and set outliers furthur than that to that value to reduce their impact on the data when doing analysis.